lib/class-wp-style-engine-gutenberg.php

```diff
@@ -0,0 +1,116 @@
+<?php
+/**
+ * WP_Style_Engine class
+ *
+ * @package Gutenberg
+ */
+
+/**
+ * Singleton class representing the style engine.
+ *
+ * Consolidates rendering block styles to reduce duplication and streamline
+ * CSS styles generation.
+ *
+ * @since 6.0.0
+ */
+class WP_Style_Engine_Gutenberg {
+	/**
+	 * Registered CSS styles.
+	 *
+	 * @since 5.5.0
+	 * @var array
+	 */
+	private $registered_styles = array();
+
+	/**
+	 * Container for the main instance of the class.
+	 *
+	 * @since 5.5.0
+	 * @var WP_Style_Engine_Gutenberg|null
+	 */
+	private static $instance = null;
+
+	/**
+	 * Register action for outputting styles when the class is constructed.
+	 */
+	public function __construct() {
+		// Borrows the logic from `gutenberg_enqueue_block_support_styles`.
+		// $action_hook_name = 'wp_footer';
+		// if ( wp_is_block_theme() ) {
+		// 	$action_hook_name = 'wp_enqueue_scripts';
+		// }
+		// add_action(
+		// 	$action_hook_name,
+		// 	array( $this, 'output_styles' )
+		// );
+	}
+
+	/**
+	 * Utility method to retrieve the main instance of the class.
+	 *
+	 * The instance will be created if it does not exist yet.
+	 *
+	 * @return WP_Style_Engine_Gutenberg The main instance.
+	 */
+	public static function get_instance() {
+		if ( null === self::$instance ) {
+			self::$instance = new self();
+		}
+
+		return self::$instance;
+	}
+
+	public function reset() {
+		$this->registered_styles = array();
+	}
+
+	public function get_generated_styles() {
+		$output = '';
+		foreach ( $this->registered_styles as $selector => $rules ) {
+			$output .= "{$selector} {\n";
+
+			if ( is_string( $rules ) ) {
+				$output .= '  ';
+				$output .= $rules;
+			} else {
+				foreach ( $rules as $rule => $value ) {
+					$output .= "  {$rule}: {$value};\n";
+				}
+			}
+			$output .= "}\n";
+		}
+		return $output;
+	}
+
+	/**
+	 * Stores style rules for a given CSS selector (the key) and returns an associated classname.
+	 *
+	 * @param string $key     A class name used to construct a key.
+	 * @param array  $options An array of options, rules, and selector for constructing the rules.
+	 *
+	 * @return string The class name for the added style.
+	 */
+	public function add_style( $key, $options ) {
```

Comment:
Ok, so If I'm understanding properly, the style engine here has a different meaning than the style engine package in JS.

The style engine package in JS is responsible for "style object -> css rules" transformation (potentially also css rules -> css string) while the current php style engine is more for "css rules -> css string". Please correct me if I'm wrong.

I kind of feel like the first step is actually the most important for the style engine work (style object -> css rule) and the second step is an optimization step (array of css rules -> css string).

That said, that's fine, if we're focusing for the second part of the style engine, I'm wondering why are we building the API like that:

php
```
$engine->add_style( $rule );
$engine->add_style( $rule );
$engine->add_style( $rule );

$css = $engine-> output_styles();
```

instead of just

```
$css = engine_output_styles( $rules );
```

the latter being stateless and also more aligned with what I think will be the output of the first step of the style engine (an array of rules). 

I guess the proposed API is motivated by the fact that the current block supports are "discrete" php files but personally I feel like they shouldn't be discrete, it's just the way they were implemented originally. I guess we can try to discuss this.

Reply:
Thanks for checking this out!

I'll let @andrewserong elaborate with specific reference to this PR, but I have some general thoughts:

> I kind of feel like the first step is actually the most important for the style engine work (style object -> css rule) and the second step is an optimization step (array of css rules -> css string).

Agree. The first step is happening over at:

https://github.com/WordPress/gutenberg/pull/39446 👍 

I see this following the stateless approach you mention.

This PR (and https://github.com/WordPress/gutenberg/pull/38974) is part of our explorations into what's possible and what we (think we) want to see down the track as well (CSS consolidation, optimisation and a better developer experience).

I think we also wanted to noncommittally test some of the hypotheses/feature requests that have been floating around, e.g., 

- obfuscation 
- consisten css classname hooks
- deduping

Layout was a good starting point for this given the amount of CSS we generate on the frontend and rule duplication. 

> I guess the question here is what is $resitered_styles for?

I believe it stems from the long-term idea that we could render all CSS in one style block on the frontend (instead of many), and, in doing so, leave open the possibility for pre-render processing/filtering of CSS rules.

Reply:
Thanks for the questions @youknowriad! I'll answer slightly out of order to hopefully explain some of my thinking 🙂

At the outset, I do want to flag that in the current state, this PR is _weird_ and experimental! 😁 My approach at the moment is to stuff this PR full of ideas / hack things in there to get them working, and then gradually pull things out into something neater.

Just echoing what @ramonjd wrote above, but the goal of this PR was to take a bit of a further out look to see what kind of end point we'll need to reach if we want to achieve some of the things we're hoping to get out of the style engine work. In Ramon's PR (#39446), the stateless / closer to the JS package version is being explored, and that'll likely be in good shape to merge much sooner than this PR.

> I kind of feel like the first step is actually the most important for the style engine work (style object -> css rule) and the second step is an optimization step (array of css rules -> css string).

Absolutely. But I think that optimisation step could be harder to implement if we don't design for it — I'm keen for us to improve the style output (primarily de-duping and consolidating to a single stylesheet), and we get quite a bit for free if we add in the internal state and separate out the two steps (adding styles and output).

I think being able to separate generating and outputting styles into separate steps via some internal state will probably allow us some good flexibility for doing things like:

* De-duping styles
* Sharing classnames or styles across functions / at different execution times
* Being able to access generated styles separately — for example, if we want to explore at some point making generated styles available via an API endpoint, having easy access to that state will make those explorations much easier

While those ideas sound very "optimise-ey" — i.e. things we'd usually think to do in later steps, and not do prematurely — here, we've got some concrete optimisations that we know we need to make and solve for, so I think it's worth it to dig in a bit to see what might look viable.

> I guess the proposed API is motivated by the fact that the current block supports are "discrete" php files but personally I feel like they shouldn't be discrete, it's just the way they were implemented originally. I guess we can try to discuss this.

It's less motivated by the separate PHP files, and more motivated by styles being output one block at a time as each block is rendered, resulting in one `style` tag per rendered block (that uses the `container` id approach). We currently abstract that away in a small way in [gutenberg_enqueue_block_support_styles](https://github.com/wordpress/gutenberg/blob/trunk/lib/compat/wordpress-5.9/script-loader.php#L63) — but my hunch at the moment is that we'll benefit from having some more smarts around this rather than relying on the `wp_footer` and `wp_head` hooks directly.

However, this is all still very much experimental! I'm more than happy for us to see how far we can get with the stateless approach in the shorter-term — my main interest right now is to ensure that if we _do_ want to add in the stateful approach it's easy enough for us to migrate it in there to address optimisation, rather than having to rewrite everything! So, at least from my perspective, the current approach has pointed toward using a class for the style engine so that we can encapsulate (and namespace) our logic, and make it easy to add in the internal state if / when we need it.

Reply:
Thanks that make sense.

> It's less motivated by the separate PHP files, and more motivated by styles being output one block at a time as each block is rendered, resulting in one style tag per rendered block (that uses the container id approach). 

That makes sense but personally what I don't like is the singleton because it's very hard to reason about depending on where you are (regular php template, api...). I'm just nitpicking at this point but I'd have a small preference for something more explicit.

Basically, what I don't like about the singleton approach (and I guess all regular WP style php) is that it's all dependent on order of execution of things, actions... It's not really clear when are we allowed to add styles to the instance, from where and who can do it. So we end up with hidden bugs, for instance what happens if I add a style to the singleton after the `generate` function has been called?

My "weird" mind prefers something more dumb :P
 
```
function render_blocks( $blocks ) {
   $markup = '';
   $rules = array();
   foreach( $blocks as $block ) {
      list( $css_rules, $block_markup ) = render_block( $block );
      $markup .= $block_markup;
      $rules = array_concat( $rules, $css_rules );
   }

   return [ $rules, $markup ];
}

list( $rules, $markup ) = render_blocks( $my_block_list );
$css = generate_optimized_stylesheet( $rules );
```

I guess it might not be easy to implement because it kinds of touches the fundamental of the existing `do_blocks` but anyway it's worth sharing just in case 🤷

we can also imagine that later we could also return JS in addition to "css_rules" and "markup" if we want to do JS optimizations.

```
list( $css_rules, $block_markup, $scripts ) = render_block( $block );
```

---
Wanted to clarify that there's no blockers from me here, I'm just basically sharing my mental model in case it helps in any way.

Reply:
> I'm just basically sharing my mental model in case it helps in any way.

Thank you, it's always very much appreciated! 😊 That code has a more JS-ey flavour to it that immediately feels more at home to me, too.

You make great points about the challenges with singletons, actions, and the order of execution. I hadn't even thought of seeing if we could mess with the `render_block` API 😄 — while changing the return type there could open up a can of worms, this is some very helpful nudging for us to try to keep the API we design explicit and hopefully simple to work with. I'll continue having a play!

Reply:
> Basically, what I don't like about the singleton approach (and I guess all regular WP style php) is that it's all dependent on order of execution of things, actions... It's not really clear when are we allowed to add styles to the instance, from where and who can do it. So we end up with hidden bugs, for instance what happens if I add a style to the singleton after the generate function has been called?

Hmm, valid point, and the example is very helpful!

I suppose we haven't really decided _where_ we'd dump the contents of every registered style 😄  It's a good excuse to dig into `render_block` too to test the idea.